cpp/cmake_modules/SetupCxxFlags.cmake

```diff
@@ -604,76 +604,42 @@ endif()
 # For all builds:
 # For CMAKE_BUILD_TYPE=Debug
 #   -ggdb: Enable gdb debugging
-# For CMAKE_BUILD_TYPE=FastDebug
-#   Same as Debug, except with some optimizations on.
 # For CMAKE_BUILD_TYPE=Release
-#   -O2: Enable all compiler optimizations
+#   -O2 (not -O3): Enable compiler optimizations
 #   Debug symbols are stripped for reduced binary size.
 # For CMAKE_BUILD_TYPE=RelWithDebInfo
 #   Same as Release, except with debug symbols enabled.
 
 if(NOT MSVC)
-  string(REPLACE "-O3" "" CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE}")
-  string(REPLACE "-O3" "" CMAKE_CXX_FLAGS_RELWITHDEBINFO
-                 "${CMAKE_CXX_FLAGS_RELWITHDEBINFO}")
-
-  set(RELEASE_FLAGS "-O2 -DNDEBUG")
+  set(C_RELEASE_FLAGS "")
+  if(CMAKE_C_FLAGS_RELEASE MATCHES "-O3")
+    string(APPEND C_RELEASE_FLAGS " -O2")
```

Comment:
Hi, kou, does this patch means, when we use release, the default build is "-O3 -DNDEBUG -O2", would it use O2 to final build it rather than O3? @kou

Reply:
Personally I think it should be `if not`?

Reply:
Yes.
We don't have any benefit with `-O3`.
See also: #32703 and #13661

> -O3 can produce longer build times and larger binaries for next to no performance benefits.

Reply:
I think you're right, but here I think that, why no 

```
if (CMAKE_CXX_FLAGS_RELEASE NOT MATCHES "-O3")
  string(APPEND CXX_RELEASE_FLAGS " -O2")
endif()
```

If I want to test O3, maybe we should have ability to use it?

Reply:
If this uses the condition, `-O3 -DNDEBUG` isn't change. We want to use `-O3 -DNDEBUG -O2` here.

We need to comment out the code when we try `-O3`. Sorry. We may want to add an option to force using `-O3`.

Reply:
Thanks, the code and comment here is clear. I just want to make clear why we we O2, because when compiling I found the argument is "-O3 -NDEBUG -O2", which make me confused.

I'll try some benchmark under x86, if I got benefits when enabling O3, I'd like to add a flag here.